tirocks/src/error.rs

```diff
@@ -0,0 +1,206 @@
+    }
+
+    #[inline]
+    pub fn with_space_limit(msg: impl AsRef<[u8]>) -> Status {
+        Self::new(Code::kIOError, SubCode::kSpaceLimit, msg)
+    }
+
+    #[inline]
+    pub fn with_path_not_found(msg: impl AsRef<[u8]>) -> Status {
+        Self::new(Code::kIOError, SubCode::kPathNotFound, msg)
+    }
+
+    #[inline]
+    pub fn ok(&self) -> bool {
+        self.0.ok()
+    }
+
+    #[inline]
+    pub fn state(&self) -> Option<&[u8]> {
+        self.0.state()
+    }
+
+    #[inline]
+    pub fn message(&self) -> std::result::Result<Option<&str>, Utf8Error> {
+        self.0.message()
+    }
+
+    #[inline]
+    pub fn code(&self) -> Code {
+        self.0.code()
+    }
+
+    #[inline]
+    pub fn severity(&self) -> Severity {
+        self.0.severity()
+    }
+
+    #[inline]
+    pub fn set_severity(&mut self, severity: Severity) {
+        self.0.set_severity(severity);
+    }
+
+    #[inline]
+    pub fn sub_code(&self) -> SubCode {
+        self.0.sub_code()
+    }
+
+    #[inline]
+    pub fn set_sub_code(&mut self, sub_code: SubCode) {
+        self.0.set_sub_code(sub_code)
+    }
+
+    #[inline]
+    pub(crate) fn into_raw(self) -> rocksdb_Status {
+        self.0
+    }
+
+    #[inline]
+    pub fn path_not_exist(&self) -> bool {
+        self.code() == Code::kNotFound
+            || self.code() == Code::kIOError && self.sub_code() == SubCode::kPathNotFound
+    }
+}
+
+impl From<rocksdb_Status> for Status {
+    #[inline]
+    fn from(s: rocksdb_Status) -> Status {
+        Status(s)
+    }
+}
+
+impl From<Status> for io::Error {
+    #[inline]
+    fn from(s: Status) -> Self {
+        let kind = match s.code() {
+            Code::kIOError => match s.sub_code() {
+                #[cfg(feature = "nightly")]
+                SubCode::kNoSpace => io::ErrorKind::StorageFull,
+                #[cfg(feature = "nightly")]
+                SubCode::kSpaceLimit => io::ErrorKind::FilesystemQuotaExceeded,
+                SubCode::kMemoryLimit => io::ErrorKind::OutOfMemory,
+                SubCode::kPathNotFound => io::ErrorKind::NotFound,
+                _ => io::ErrorKind::Other,
+            },
+            Code::kNotFound => io::ErrorKind::NotFound,
+            _ => io::ErrorKind::Other,
+        };
+        let msg = match s.state() {
+            Some(s) => String::from_utf8_lossy(s),
+            None => return Self::from(kind),
+        };
+        Self::new(kind, msg)
+    }
+}
+
+impl Debug for Status {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.ok() {
+            write!(f, "OK")
+        } else {
+            write!(
+                f,
+                "Code: {:?}, SubCode: {:?}, Severity: {:?}, msg: {:?}",
+                self.code(),
+                self.sub_code(),
+                self.severity(),
+                self.message()
+            )
+        }
+    }
+}
+
+pub type Result<T> = std::result::Result<T, Status>;
```

Comment:
Could `ffi_try` be modified to yield a result directly? E.g. `ffi_try!(f(db))?`.

Reply:
`ffi_try` already returns result directly. So it's `ffi_try!(f(db))` vs `ffi_try!(f(db))?`, why do you think write additional `?` is better?

Reply:
It follows the same design pattern of the rust builtin deprecated macros `try`, which is also the reason why it's called `ffi_try`. `try` macro will exit the function implicitly. I don't have strong opinion on whether choose exit implicitly or return a result type, but it certainly should be renamed without `try` in the latter case.

Reply:
Rename looks fine. Please replace it. It reduces some boilerplate code and is more flexible.